### ISSUE TITLE
[State Encryption] Fails to save state to Gitlab HTTP State Backend When Encrypted

### ISSUE THREAD

**Issue by tagur87** [opentofu/opentofu#1407]
### OpenTofu Version

```shell
OpenTofu v1.7.0-alpha1
on darwin_amd64
+ provider registry.opentofu.org/e-breuninger/netbox v3.8.4
```

### OpenTofu Configuration Files

```hcl
terraform {
  required_providers {
  ...
  }
  backend "http" {
    address = "https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted"
    lock_address = "https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted/lock"
    unlock_address = "https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted/lock"
    username = "user"
    password = "glpat-xyz"
    lock_method = "POST"
    unlock_method = "DELETE"
    retry_wait_min = "5"

    
  }
  encryption {
    key_provider "pbkdf2" "test" {
      passphrase = "abcd12345xyz!@#$"
      key_length = 32
      iterations = 600000
      salt_length = 32
      hash_function = "sha512"
    }

    method "aes_gcm" "encrypt" {
      keys = key_provider.pbkdf2.test
    }


    state {
      method = method.aes_gcm.encrypt
      fallback {
       
      }
    }
  }
}
```

### Debug Output

```
2024-03-15T10:44:12.456-0400 [TRACE] Executing graph transform *tofu.TransitiveReductionTransformer
2024-03-15T10:44:12.456-0400 [TRACE] Completed graph transform *tofu.TransitiveReductionTransformer (no changes)
2024-03-15T10:44:12.456-0400 [DEBUG] Starting graph walk: walkApply
2024-03-15T10:44:12.456-0400 [TRACE] vertex "root": starting visit (*tofu.nodeCloseModule)
2024-03-15T10:44:12.456-0400 [TRACE] vertex "root": visit complete
2024-03-15T10:44:12.456-0400 [DEBUG] states/remote: state read serial is: 1; serial is: 1
2024-03-15T10:44:12.456-0400 [DEBUG] states/remote: state read lineage is: 4ddf5053-a993-fc23-eda1-2383a5767d64; lineage is: 4ddf5053-a993-fc23-eda1-2383a5767d64
2024-03-15T10:44:12.802-0400 [DEBUG] POST https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted?ID=df7262f3-793e-9689-b010-e263a5b3d4db
2024-03-15T10:44:12.958-0400 [DEBUG] POST https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted?ID=df7262f3-793e-9689-b010-e263a5b3d4db (status: 500): retrying in 5s (2 left)
2024-03-15T10:44:18.087-0400 [DEBUG] POST https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted?ID=df7262f3-793e-9689-b010-e263a5b3d4db (status: 500): retrying in 10s (1 left)
2024-03-15T10:44:28.226-0400 [TRACE] statemgr.Filesystem: reading initial snapshot from errored.tfstate
2024-03-15T10:44:28.926-0400 [TRACE] statemgr.Filesystem: read snapshot with lineage "4ddf5053-a993-fc23-eda1-2383a5767d64" serial 3
2024-03-15T10:44:28.926-0400 [TRACE] statemgr.Filesystem: Importing snapshot with lineage "4ddf5053-a993-fc23-eda1-2383a5767d64" serial 2 over snapshot with lineage "4ddf5053-a993-fc23-eda1-2383a5767d64" serial 3 at errored.tfstate
2024-03-15T10:44:28.926-0400 [TRACE] statemgr.Filesystem: forcing lineage "4ddf5053-a993-fc23-eda1-2383a5767d64" serial 2 for migration/import
2024-03-15T10:44:28.926-0400 [TRACE] statemgr.Filesystem: preparing to manage state snapshots at errored.tfstate
2024-03-15T10:44:29.603-0400 [TRACE] statemgr.Filesystem: existing snapshot has lineage "4ddf5053-a993-fc23-eda1-2383a5767d64" serial 3
2024-03-15T10:44:29.603-0400 [TRACE] statemgr.Filesystem: state file backups are disabled
2024-03-15T10:44:29.603-0400 [TRACE] statemgr.Filesystem: no state changes since last snapshot
2024-03-15T10:44:29.603-0400 [TRACE] statemgr.Filesystem: writing snapshot at errored.tfstate
╷
│ Error: Failed to save state
│ 
│ Error saving state: Failed to upload state: POST https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted?ID=df7262f3-793e-9689-b010-e263a5b3d4db giving up after 3 attempt(s)
╵
╷
│ Error: Failed to persist state to backend
│ 
│ The error shown above has prevented OpenTofu from writing the updated state to the configured backend. To allow for recovery, the state has been written to the file "errored.tfstate" in the current working directory.
│ 
│ Running "tofu apply" again at this point will create a forked state, making it harder to recover.
│ 
│ To retry writing this state, use the following command:
│     tofu state push errored.tfstate
│ 
╵
2024-03-15T10:44:29.948-0400 [DEBUG] DELETE https://gitlab.com/api/v4/projects/1234/terraform/state/encrypted/lock

```

### Expected Behavior

HTTP Backend should work with the encrypted state format.

### Actual Behavior

HTTP 500 Error when saving state to Gitlab HTTP state backend.

### Steps to Reproduce

1. Configure HTTP backend config
2. `tofu init`
3. `tofu apply`
4. Change config to add encryption
5. `tofu apply`

**Comment by cam72cam** [comment]
The gitlab http remote state service will need to be modified to support encrypted state: https://gitlab.com/gitlab-org/gitlab/-/blob/id-rails-main/app/models/terraform/state.rb?ref_type=heads#L78

**Comment by janosdebugs** [comment]
@cam72cam it looks like the data needed by Gitlab may be something we could expose in the encrypted version, but lets wait for more feedback if other systems require similar things.

@tagur87 thank you so much for your help in testing this release!

**Comment by cam72cam** [comment]
Cross linking to the Gitlab issue: https://gitlab.com/gitlab-org/gitlab/-/issues/450816

**Comment by janosdebugs** [comment]
@cam72cam so looking at the code you referenced in Gitlab, the comment says that it can actually be removed already.

**Comment by janosdebugs** [comment]
We have spoken with @cam72cam and we are considering exposing the following metadata fields:

- version (currently: 4)
- terraform_version (currently: 1.7.0)
- lineage (currently: UUID that stays the same for the project)
- serial (serial number, essentially the version of the state file)

However, we have some concerns for both exposing information that could be used to attack the encrypted version, and we are also concerned that it may look like a valid state file and someone could mistakenly run Terraform on it, causing some havoc.

My personal feeling is that lineage and serial may be safe to expose.

**Comment by cam72cam** [comment]
I have a similar feeling.  I don't think we should expose the state version ("version") field as that is generally used in the initial setps of identifying a valid terraform/tofu state file.  The rest may already be used by the http backends (such as gitlab).

**Comment by StephanHCB** [comment]
This is just guesswork on my part, but I agree that the most important risk is that tooling might mistake encrypted state for ordinary state and then do weird things, especially with the way many json unmarshal implementation just skip unknown fields.

From a known-plaintext attack perspective: If there is a known plaintext attack against the method, then we've lost anyway, as there is already a known part of the plaintext of any state encountered in the wild (they all start with a pretty-printed version of `{"version":4,"terraform_version":`) and adding a few more bytes to this very predictable text at a more randomized location may not make such a big difference.

**Comment by cam72cam** [comment]
I'm going to update my PR to include lineage field and get it merged so this makes it into 1.7.0-alpha2

CC @timofurrer
